libyara/modules/macho.c

```diff
@@ -0,0 +1,1225 @@
+      set_integer(yr_##bo##32toh(sec_64->reserved3),                           \
+                  object, "segments[%i].sections[%i].reserved3", i, j);        \
+    }                                                                          \
+  }                                                                            \
+}                                                                              \
+
+MACHO_HANDLE_SEGMENTS_SECTIONS(32,le)
+MACHO_HANDLE_SEGMENTS_SECTIONS(64,le)
+MACHO_HANDLE_SEGMENTS_SECTIONS(32,be)
+MACHO_HANDLE_SEGMENTS_SECTIONS(64,be)
+
+
+// Parse Mach-O file with specific bit-width and byte order.
+
+#define MACHO_PARSE_FILE(bits,bo)                                              \
+void macho_parse_file_##bits##_##bo(                                           \
+    void* data,                                                                \
+    uint64_t size,                                                             \
+    YR_OBJECT* object,                                                         \
+    YR_SCAN_CONTEXT* context)                                                  \
+{                                                                              \
+  mach_header_##bits##_t* header = (mach_header_##bits##_t*)data;              \
+                                                                               \
+  if (size < sizeof(mach_header_##bits##_t))                                   \
+    return;                                                                    \
+                                                                               \
+  set_integer(yr_##bo##32toh(header->magic), object, "magic");                 \
+  set_integer(yr_##bo##32toh(header->cputype), object, "cputype");             \
+  set_integer(yr_##bo##32toh(header->cpusubtype), object, "cpusubtype");       \
+  set_integer(yr_##bo##32toh(header->filetype), object, "filetype");           \
+  set_integer(yr_##bo##32toh(header->ncmds), object, "ncmds");                 \
+  set_integer(yr_##bo##32toh(header->sizeofcmds), object, "sizeofcmds");       \
+  set_integer(yr_##bo##32toh(header->flags), object, "flags");                 \
+  if (bits == 64)                                                              \
+  {                                                                            \
+    mach_header_64_t* header_64 = (mach_header_64_t*)data;                     \
+    set_integer(yr_##bo##32toh(header_64->reserved), object, "reserved");      \
+  }                                                                            \
+                                                                               \
+  uint64_t num_sg = 0;                                                         \
+  uint8_t *command = (uint8_t*)(header + 1);                                   \
+  for (unsigned i = 0; i < yr_##bo##32toh(header->ncmds); i++)                 \
```

Comment:
No sanity checks for preventing out-of-bounds reads. If `header->ncmds` is altered or the file is truncated after the header this will segfault.